src/senaite/core/upgrade/v02_04_000.py

```diff
@@ -137,3 +139,46 @@ def import_typeinfo(tool):
     setup = portal.portal_setup
     setup.runImportStepFromProfile("profile-bika.lims:default", "typeinfo")
     setup.runImportStepFromProfile(profile, "typeinfo")
+
+
+def fix_traceback_retract_dl(tool):
+    """Migrates the values of LDL and UDL of analyses/services to string, as
+    well as results that are DetectionLimit and stored as floats
+    """
+    logger.info("Migrate LDL, UDL and result fields to string ...")
+    cat = api.get_tool("uid_catalog")
+    query = {"portal_type": ["AnalysisService", "Analysis",
+                             "DuplicateAnalysis", "ReferenceAnalysis"]}
+    brains = cat.search(query)
+    total = len(brains)
+
+    for num, brain in enumerate(brains):
+        if num and num % 100 == 0:
+            logger.info("Migrated {0}/{1} LDL/UDL fields".format(num, total))
+
+        obj = api.get_object(brain)
+
+        # Migrate UDL to string
+        field = obj.getField("UpperDetectionLimit")
+        value = field.get(obj)
+        if isinstance(value, (int, float)):
+            field.set(obj, str(value))
+
+        # Migrate LDL to string
+        field = obj.getField("LowerDetectionLimit")
+        value = field.get(obj)
+        if isinstance(value, (int, float)):
+            field.set(obj, str(value))
+
+        # Migrate the result
+        field = obj.getField("Result")
+        if field and obj.getDetectionLimitOperand() in [LDL, UDL]:
+            # The result is the detection limit
+            result = field.get(obj)
+            if isinstance(result, (int, float)):
+                field.set(obj, str(result))
+
+        # Flush the object from memory
+        obj._p_deactivate()
```

Comment:
We should add every 1000 iterations an `transaction.savepoint()` to reduce the memory size of the transaction